src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java

```diff
@@ -0,0 +1,134 @@
+        super(syncOptions);
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Optional<String>> fetchCachedCustomObjectId(
+            @Nonnull final CustomObjectCompositeIdentifier identifier) {
+
+        String container = identifier.getContainer();
+        String key = identifier.getKey();
+
+        if (StringUtils.isEmpty(container) || StringUtils.isEmpty(key)) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+
+        return fetchCachedResourceId(identifier.toString(),
+            draft -> CustomObjectCompositeIdentifier.of(draft).toString(),
+            () -> CustomObjectQuery.ofJsonNode()
+                        .withPredicates(q -> q.container().is(container).and(q.key().is(key)))
+        );
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Set<CustomObject<JsonNode>>> fetchMatchingCustomObject(
+            @Nonnull final Set<CustomObjectCompositeIdentifier> identifiers) {
+
+        Set<CustomObjectCompositeIdentifier> filteredIdentifiers = identifiers.stream()
+                .filter(identifier ->
+                            StringUtils.isNotEmpty(identifier.getContainer())
+                        &&  StringUtils.isNotEmpty(identifier.getKey()))
+                .collect(Collectors.toSet());
+
+        if (filteredIdentifiers.size() == 0) {
+            return CompletableFuture.completedFuture(Collections.EMPTY_SET);
+        }
+
+        Set<String> identifierStrings =
+                identifiers.stream().map(CustomObjectCompositeIdentifier::toString).collect(Collectors.toSet());
+
+        return fetchMatchingResources(identifierStrings,
+            draft -> CustomObjectCompositeIdentifier.of(draft).toString(),
+            () -> {
+                CustomObjectQueryBuilder<JsonNode> query = CustomObjectQueryBuilder.ofJsonNode();
+                query = query.plusPredicates(queryModel -> createQuery(queryModel, identifiers));
+                return query.build();
+            });
+    }
+
+    @Nonnull
+    private QueryPredicate<CustomObject<JsonNode>> createQuery(
+            @Nonnull final CustomObjectQueryModel<CustomObject<JsonNode>> queryModel,
+            @Nonnull final Set<CustomObjectCompositeIdentifier> identifiers) {
+
+        QueryPredicate<CustomObject<JsonNode>> queryPredicate = QueryPredicate.of(null);
+        boolean firstAttempt = true;
+        for (CustomObjectCompositeIdentifier identifier : identifiers) {
+            String key = identifier.getKey();
+            String container = identifier.getContainer();
+            if (firstAttempt) {
+                queryPredicate = queryModel.container().is(container).and(queryModel.key().is(key));
+                firstAttempt = false;
+            } else {
+                queryPredicate = queryPredicate.or(queryModel.container().is(container).and(queryModel.key().is(key)));
+            }
+        }
+        return queryPredicate;
```

Comment:
This method can be reduced to one reduce. Check if it works:

```suggestion
        return identifiers.stream()
                          .reduce(QueryPredicate.of(null),
                                  (partialQuery, identifier) -> {
                                      String key = identifier.getKey();
                                      String container = identifier.getContainer();
                                      return queryModel.container().is(container).and(queryModel.key().is(key));
                                  }, QueryPredicate::or);
```

Reply:
I added one integration test to test it but failed. So I keep the original one.
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/9822feafbd5d2be063b25bbfea83ae2684383ff0